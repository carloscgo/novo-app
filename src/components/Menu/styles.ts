import styled from 'styled-components';
import { Link } from "react-router-dom";
import { NavDropdown } from 'react-bootstrap';

const Container = styled.div`
  width: 100%;
  height: 25px;
  background-color: var(--bg-white);
  display: flex;
  align-items: center;

  .item-width {
    min-width: 120px;
  }
`

const LinkMenu = styled(Link)`
  text-decoration: none !important;
  color: var(--text-black);
  border-bottom: 3px solid transparent;
  padding-bottom: 10px;

  &:hover {
    color: var(--text-black);
    border-bottom: 2px solid var(--main-color);
  }
`

const DropdownMenu = styled(NavDropdown)`
  text-decoration: none !important;
  color: var(--text-black);
  border-bottom: 2px solid transparent;
  padding-bottom: 10px;
  
  &:hover .dropdown-menu {
    display: block;
  }

  .dropdown-menu > div {
    flex-direction: column;
    height: auto;
  }

  &:hover {
    color: var(--text-black);
    border-bottom: 2px solid var(--main-color);
  }

  a {
    color: var(--text-black);
    
      &:hover {
      color: var(--text-black);
    }
  }

  .dropdown-toggle {
    padding-bottom: 0 !important;

    &::after {
      border-top: none !important;
    }
  }
`

export {
  Container,
  LinkMenu,
  DropdownMenu,
}