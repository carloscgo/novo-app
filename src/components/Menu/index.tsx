import Nav from 'react-bootstrap/Nav';
import isArray from 'lodash/isArray';
import {
  CProps, TItem
} from '../../utils/interfaces';
import { LinkMenu, DropdownMenu, Container } from './styles';

const Menu = ({ items, onClick, selection }: CProps) => {
  const hasChildren = (item: TItem) => isArray(item.children) && !!item.children.length

  return (
    <Container>
      {isArray(items) && items.map((item: TItem) =>
        <Nav key={item.id} className={`item-width align-items-start ` + (item.id === selection?.id ? 'selected' : null) as string}>
          {hasChildren(item) && (
            <DropdownMenu title={item.name} id={`${item.id}-nav-dropdown`} renderMenuOnMount={true}>
              <Menu
                items={item.children}
                onClick={onClick}
                selection={selection}
              />
            </DropdownMenu>
          )}

          {!hasChildren(item) && (
            <Nav.Link as='div'>
              <LinkMenu to="#" aria-label={`link-menu-${item.id}`} onClick={() => onClick(item)}>
                {item.name}
              </LinkMenu>
            </Nav.Link>
          )}
        </Nav>
      )}
    </Container>
  )
}

export default Menu