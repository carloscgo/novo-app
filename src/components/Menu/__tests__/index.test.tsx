/**
 *
 * Tests for Menu
 *
 */

import Menu from '../index'

import { render, BrowserRouter, fireEvent, act } from '../../../../jest.setup'

describe('<Menu />', () => {
  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(() => { })
  })

  const onClick = jest.fn()

  const items = [{
    id: '1',
    name: 'item 1'
  }, {
    id: '2',
    name: 'item 2',
    children: [{
      id: '3',
      name: 'item 3'
    }]
  }]

  const renderComponent = (items: any[], selection: any = null) => render(
    <BrowserRouter>
      <Menu items={items} onClick={onClick} selection={selection} />
    </BrowserRouter>
  )

  it('expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error')

    renderComponent([])

    expect(spy).not.toHaveBeenCalled()
  })

  it('expect render with items', () => {
    const spy = jest.spyOn(global.console, 'error')

    renderComponent(items)

    expect(spy).not.toHaveBeenCalled()
  })

  it('expect render with onClick event', () => {
    const container = renderComponent(items)

    const onClick = container.getByLabelText('link-menu-1')

    fireEvent.click(onClick)

    expect(container.baseElement).toMatchSnapshot()
  })

  it('expect render with selected item', () => {
    const container = renderComponent(items, items[0])

    expect(container.baseElement).toMatchSnapshot()
  })
})
