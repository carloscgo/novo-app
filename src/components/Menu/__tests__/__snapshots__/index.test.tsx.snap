// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> expect render with onClick event 1`] = `
<body>
  <div>
    <div
      class="sc-bcXHqe jzLKGg"
    >
      <div
        class="item-width align-items-start null nav"
      >
        <div
          class="nav-link"
        >
          <a
            aria-label="link-menu-1"
            class="sc-gswNZR cEpWoe"
            href="/"
          >
            item 1
          </a>
        </div>
      </div>
      <div
        class="item-width align-items-start null nav"
      >
        <div
          class="sc-dkrFOg ffwown nav-item dropdown"
        >
          <a
            aria-expanded="false"
            class="dropdown-toggle nav-link"
            href="#"
            id="2-nav-dropdown"
            role="button"
            tabindex="0"
          >
            item 2
          </a>
          <div
            aria-labelledby="2-nav-dropdown"
            class="dropdown-menu"
            x-placement="bottom-start"
          >
            <div
              class="sc-bcXHqe jzLKGg"
            >
              <div
                class="item-width align-items-start null nav"
              >
                <div
                  class="nav-link"
                >
                  <a
                    aria-label="link-menu-3"
                    class="sc-gswNZR cEpWoe"
                    href="/"
                  >
                    item 3
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`<Menu /> expect render with selected item 1`] = `
<body>
  <div>
    <div
      class="sc-bcXHqe jzLKGg"
    >
      <div
        class="item-width align-items-start selected nav"
      >
        <div
          class="nav-link"
        >
          <a
            aria-label="link-menu-1"
            class="sc-gswNZR cEpWoe"
            href="/"
          >
            item 1
          </a>
        </div>
      </div>
      <div
        class="item-width align-items-start null nav"
      >
        <div
          class="sc-dkrFOg ffwown nav-item dropdown"
        >
          <a
            aria-expanded="false"
            class="dropdown-toggle nav-link"
            href="#"
            id="2-nav-dropdown"
            role="button"
            tabindex="0"
          >
            item 2
          </a>
          <div
            aria-labelledby="2-nav-dropdown"
            class="dropdown-menu"
            x-placement="bottom-start"
          >
            <div
              class="sc-bcXHqe jzLKGg"
            >
              <div
                class="item-width align-items-start null nav"
              >
                <div
                  class="nav-link"
                >
                  <a
                    aria-label="link-menu-3"
                    class="sc-gswNZR cEpWoe"
                    href="/"
                  >
                    item 3
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
